### DISCUSSION TITLE
! [Target] Command "zfs list <target dataset>" returned exit code 127 (valid codes: [0, 1])

### DISCUSSION THREAD

**Discussion by Zamana** [psy0rz/zfs_autobackup#167] [General]
Hi!

I created a script to backup my datasets, but it behaves differently, according to the way I run it:

1) When I run it through the command line, it runs fine;

but

2) When I run it through crontab, it always gives this same message on each and every dataset:

```
Selecting dataset property : autobackup:ebooks
  Snapshot format            : ebooks-%Y%m%d%H%M%S
  Hold name                  : zfs_autobackup:ebooks

  #### Source settings
  [Source] Datasets on: pve
  [Source] Keep the last 3 snapshots.

  #### Selecting
  [Source] storage/ebooks: Selected

  #### Snapshotting
  [Source] Creating snapshots ebooks-20220816123003 in pool storage

  #### Target settings
  [Target] Datasets are local
  [Target] Keep the last 3 snapshots.
  [Target] Receive datasets under: storage/autobkp

  #### Synchronising
! [Target] Command "zfs list storage/autobkp" returned exit code 127 (valid codes: [0, 1])
! Exception: Last command returned error
  zfs-autobackup v3.1.3 - (c)2021 E.H.Eefting ([edwin@datux.nl](mailto:edwin@datux.nl))
```

What is that mean and what's the difference when running at the console or via crontab?

This is a PULL backup.

**Comment by psy0rz** [comment]
Does the cronjob run as the correct user ?

If you ssh to "pve" does it login as root? (Configured via ~/.ssh/config probably)

**Comment by psy0rz** [comment]
Oh never mind: A 127 is a command-not-found. This means the path to the zfs binary isnt correctly set when running from the cron.

Try using something like

`
0 0 * * * PATH=$PATH:/path/to/zfs/command/bin zfs-autobackup ....
`

**Reply by OleHolmNielsen** [reply]
This extra PATH was not sufficient on our AlmaLinux 8.7 server.  It turns out that crontab jobs don't have ```/usr/sbin``` nor ```/usr/local/bin``` in the PATH.  The installation on EL8 has ```zfs-autobackup``` in ```/usr/local/bin``` and the ```zfs``` command in ```/usr/sbin```.
The following fixes the PATH issue:
```
0 4 * * * PATH=$PATH:/usr/sbin:/usr/local/bin; zfs-autobackup -v --ssh-source <source> <property-name> <poolname>

```

**Comment by Zamana** [comment]
Exactly. That was the missing point.

Fixed.

Thank you very much @psy0rz !
